File: packages/plugin-sdk/src/events.ts
// The `events` primitive — durable subscribe over the Postgres-backed bus (H6). A
// BACKEND primitive (double token). Delivery is AT-LEAST-ONCE: each event carries
// a monotonic `id` that is the idempotency key — the caller dedupes on it. A
// subscriber that was down REPLAYS its backlog from its offset on the next poll.
// (The lite bus is Postgres; Redis Streams is the deferred scale backend behind
// the same contract — this client is unchanged across both.)
import { SdkError } from "./client.js";

export interface EventsConfig {
  baseUrl: string;
  authHeaders: () => Record<string, string>;
  fetchImpl?: typeof fetch;
}

export interface Event {
  id: number; // monotonic offset + idempotency key
  topic: string;
  project_id: string;
  subject_id: string; // fetch the entity via the query primitive
}

export class EventsClient {
  constructor(private readonly cfg: EventsConfig) {}

  private call(op: string, body: unknown): Promise<Response> {
    const f = this.cfg.fetchImpl ?? fetch;
    return f(`${this.cfg.baseUrl}/v1alpha1/plugin/events/${op}`, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...this.cfg.authHeaders() },
      body: JSON.stringify(body),
    });
  }

  /** Poll for events across topics (does NOT advance the offset — ack does). */
  async poll(topics: string[], max = 100): Promise<Event[]> {
    const res = await this.call("poll", { topics, max });
    if (!res.ok) throw await eErr(res);
    return (await res.json()).events as Event[];
  }

  /** Ack a topic up to (and including) an event id — stops re-delivery. */
  async ack(topic: string, offset: number): Promise<void> {
    const res = await this.call("ack", { topic, offset });
    if (!res.ok && res.status !== 204) throw await eErr(res);
  }

  /**
   * Dead-letter ONE event this subscriber can never process — a PERMANENT failure
   * (malformed subject, a deterministic handler rejection). The event is recorded to
   * the dead-letter log and the offset advances past it, so the good events queued
   * behind it flow on the next poll.
   *
   * Use this ONLY for permanent failures. For a TRANSIENT failure (your downstream is
   * momentarily down), do NOT ack and do NOT fail — the event is re-delivered on the
   * next poll and is never dropped. `id` must be the head of your unacked window (the
   * lowest un-acked id): ack the good events before the poison first, else the kernel
   * refuses with a 409.
   */
  async fail(topic: string, id: number, reason: string): Promise<void> {
    const res = await this.call("fail", { topic, id, reason });
    if (!res.ok && res.status !== 204) throw await eErr(res);
  }

  /**
   * Run a poll→handle→ack loop until stopped. The handler is invoked once per event
   * in id order; because delivery is at-least-once the handler MUST be idempotent (it
   * receives the event id to dedupe). Returns a stop() function.
   *
   * The handler signals the permanent-vs-transient taxonomy by how it fails:
   *   - returns normally  → success; the event is acked.
   *   - throws PermanentEventError → the event is dead-lettered (fail) and skipped;
   *     the loop continues with the events behind it.
   *   - throws anything else → treated as TRANSIENT: the batch stops at that event and
   *     it is re-delivered on the next poll (never dropped). Ack/fail happen per event,
   *     so a permanent poison never blocks the good events queued behind it.
   */
  subscribe(topics: string[], handler: (e: Event) => Promise<void> | void, opts: { intervalMs?: number } = {}): () => void {
    let stopped = false;
    const interval = opts.intervalMs ?? 1000;
    const loop = async () => {
      while (!stopped) {
        let progressed = false;
        try {
          const events = await this.poll(topics);
          for (const e of events) {
            if (stopped) break;
            try {
              await handler(e);
              await this.ack(e.topic, e.id);
              progressed = true;
            } catch (err) {
              if (err instanceof PermanentEventError) {
                // Permanent: dead-letter this one and advance past it. Because we ack
                // each event as it succeeds, this event is the head of the unacked
                // window, so fail() is accepted.
                await this.fail(e.topic, e.id, err.reason);
                progressed = true;
                continue;
              }
              // Transient: stop the batch here. The offset sits at the last acked event,
              // so this one is re-delivered on the next poll — retried, never dropped.
              break;
            }
          }
        } catch {
          /* poll failed; back off below */
        }
        if (!progressed) await sleep(interval);
      }
    };
    void loop();
    return () => {
      stopped = true;
    };
  }
}

/**
 * Throw this from a subscribe() handler to mark an event as a PERMANENT failure — one
 * that can never succeed for this subscriber (a malformed subject, a deterministic
 * rejection). The event is dead-lettered and skipped. For a transient failure, throw
 * any other error (or let one propagate) and the event is retried on the next poll.
 */
export class PermanentEventError extends Error {
  constructor(public readonly reason: string) {
    super(`permanent event failure: ${reason}`);
    this.name = "PermanentEventError";
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

async function eErr(res: Response): Promise<SdkError> {
  let code = "events_error";
  try {
    code = ((await res.json()) as { error?: string }).error ?? code;
  } catch {
    /* non-JSON */
  }
  return new SdkError(res.status, code, `events operation failed: ${code}`);
}

File: kernel/internal/gateway/pluginapi/events.go
package pluginapi

import (
	"context"
	"errors"
	"net/http"

	"github.com/m7mdhka/llmobs/kernel/internal/bus"
	"github.com/m7mdhka/llmobs/kernel/internal/gateway/pluginauth"
)

// EventBus is the subscribe surface the events endpoints need (interface-at-consumer).
type EventBus interface {
	Poll(ctx context.Context, pluginID, projectID string, topics []string, max int) ([]bus.Delivered, error)
	Ack(ctx context.Context, pluginID, projectID, topic string, upTo int64) error
	Fail(ctx context.Context, pluginID, projectID, topic string, id int64, reason string) error
}

// Events serves the `events` primitive (poll + ack), gated on cap:events. Delivery
// is at-least-once with the event id as the idempotency key; a subscriber replays
// from its offset after downtime; the tenant comes from the assertion.
type Events struct {
	authz *pluginauth.Authorizer
	bus   EventBus
}

func NewEvents(authz *pluginauth.Authorizer, b EventBus) *Events {
	return &Events{authz: authz, bus: b}
}

func (h *Events) Register(mux *http.ServeMux, prefix string) {
	mux.HandleFunc(prefix+"/poll", h.handle(h.poll))
	mux.HandleFunc(prefix+"/ack", h.handle(h.ack))
	mux.HandleFunc(prefix+"/fail", h.handle(h.fail))
}

func (h *Events) handle(fn func(http.ResponseWriter, *http.Request, pluginauth.Caller)) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}
		caller, status, err := h.authz.Require(r, "events")
		if err != nil {
			writeJSON(w, status, map[string]any{"error": err.Error()})
			return
		}
		fn(w, r, caller)
	}
}

type eventsReq struct {
	Topics []string `json:"topics,omitempty"`
	Max    int      `json:"max,omitempty"`
	Topic  string   `json:"topic,omitempty"`
	Offset int64    `json:"offset,omitempty"`
	ID     int64    `json:"id,omitempty"`
	Reason string   `json:"reason,omitempty"`
}

const maxPollTopics = 64 // per-poll topic fan-out cap (DoS bound)

func (h *Events) poll(w http.ResponseWriter, r *http.Request, c pluginauth.Caller) {
	var req eventsReq
	if err := decodeJSON(w, r, &req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "invalid JSON"})
		return
	}
	if len(req.Topics) == 0 {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "topics required"})
		return
	}
	// Cap the topic fan-out: each topic costs Offset+LatestID+After per poll, so an
	// unbounded list turns one authorized call into O(topics) backend round-trips.
	if len(req.Topics) > maxPollTopics {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "too many topics"})
		return
	}
	events, err := h.bus.Poll(r.Context(), c.PluginID, c.ProjectID, req.Topics, req.Max)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "poll failed"})
		return
	}
	if events == nil {
		events = []bus.Delivered{}
	}
	writeJSON(w, http.StatusOK, map[string]any{"events": events})
}

func (h *Events) ack(w http.ResponseWriter, r *http.Request, c pluginauth.Caller) {
	var req eventsReq
	if err := decodeJSON(w, r, &req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "invalid JSON"})
		return
	}
	if req.Topic == "" || req.Offset <= 0 {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "topic and offset required"})
		return
	}
	if err := h.bus.Ack(r.Context(), c.PluginID, c.ProjectID, req.Topic, req.Offset); err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "ack failed"})
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// fail dead-letters ONE poison event (the head of the subscriber's unacked window) and
// advances past it, so good events queued behind it flow. This is the permanent half of
// the permanent-vs-transient taxonomy: a transient failure is signalled by NOT acking
// (the event is re-delivered); a permanent one is signalled here (dead-lettered, never
// re-delivered, recorded for inspection). The tenant/plugin come from the assertion.
func (h *Events) fail(w http.ResponseWriter, r *http.Request, c pluginauth.Caller) {
	var req eventsReq
	if err := decodeJSON(w, r, &req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "invalid JSON"})
		return
	}
	if req.Topic == "" || req.ID <= 0 {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "topic and id required"})
		return
	}
	err := h.bus.Fail(r.Context(), c.PluginID, c.ProjectID, req.Topic, req.ID, req.Reason)
	if errors.Is(err, bus.ErrNotHead) {
		// The subscriber tried to fail an event ahead of its head — a client error: it
		// must ack the good events before the poison, then fail the poison. Returned as
		// a 409 so a retry-past-head is not mistaken for a server fault.
		writeJSON(w, http.StatusConflict, map[string]any{"error": "id not at head of unacked window; ack preceding events first"})
		return
	}
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "fail failed"})
		return
	}
	w.WriteHeader(http.StatusNoContent)
}
